### ISSUE TITLE
Not all code paths return a value for enums

### ISSUE THREAD

**Issue by programaths** [godotengine/godot#71763]
### Godot version

v4.0.beta13.official [caacade56]

### System information

Windows 10

### Issue description

The following code covers all execution path

```gdscript
enum Direction{UP,DOWN,LEFT,RIGHT}

func turn_left(dir)->Direction:
	match dir:
		Direction.LEFT:
			return Direction.DOWN
		Direction.RIGHT:
			return Direction.UP
		Direction.UP:
			return Direction.LEFT
		Direction.DOWN:
			return Direction.RIGHT
```
Yet it triggers "Not all code paths return a value"

### Steps to reproduce

Copy paste the following code into a node:

```gdscript
enum Direction{UP,DOWN,LEFT,RIGHT}

func turn_left(dir)->Direction:
	match dir:
		Direction.LEFT:
			return Direction.DOWN
		Direction.RIGHT:
			return Direction.UP
		Direction.UP:
			return Direction.LEFT
		Direction.DOWN:
			return Direction.RIGHT
```

### Minimal reproduction project

N/A

**Comment by KoBeWi** [comment]
Your `dir` is not typed, so it can be anything.
Although `dir: Direction` doesn't seem to help. I think enums aren't proper types.

**Comment by Vilcrow** [comment]
@programaths The function ```turn_left``` must return some value after the ```match``` statement, since the return value type is explicitly specified.

**Comment by dalexeev** [comment]
Also, even for typed code, enums are not a strict type. You can still pass in an int that is not an enum value and there will be no error.

**Comment by programaths** [comment]
@Vilcrow In Kotlin (another language), the compiler can see that all enum values are covered and will only issue an error if this not the case, telling you which values are missing. The IDE offers a quick fix to add missing cases.
```kotlin
enum class Color {
  RED, GREEN, BLUE
}

when (getColor()) {
    Color.RED -> println("red")
    Color.GREEN -> println("green")
    Color.BLUE -> println("blue")
    // 'else' is not required because all cases are covered
}

when (getColor()) {
  Color.RED -> println("red") // no branches for GREEN and BLUE
  else -> println("not red") // 'else' is required
}
```

The idea between this feature (and also in functional programming from which Kotlin took inspiration), is that if you define a default branch (or catch all branch), then you are not sure that new enum values will be handled correctly. Your program will happily compile and run with the wrong behavior.

But by being explicit, the compiler can warn you that you've to take into account the new values. A very useful feature when you come back on a codebase days or week later.

As a fix, I added "NONE" to my enum as I couldn't return null.
![image](https://user-images.githubusercontent.com/648709/213864036-4b270634-a3d2-45fc-96e3-16b3874ce0d1.png)

**Comment by akien-mga** [comment]
I'm not familiar with Kotlin but in GDScript (and in C++), enums are just integers and you could call `turn_left(42)` without error. So it needs to do something in that case.

The traditional way to handle this would be to fall back to a valid value, but also print an error (or even assert), because passing an int which is not from your enum in this case should be considered a bug from the calling code.

**Comment by programaths** [comment]
In modern languages (Java, C#) enums are indeed a bit more than integers, it has all the benefits I highlighted.
You can think of them as boolean on steroids (value set). They are useful, because you can then type properly your method and there is no way to call it with an unknown value. It also allows the compiler to check if all cases have been taken into a switch.
So, full fledged enums comes with a lot of advantages.

Enums being integers defeat a bit the type system as I could type in as receiving one enum type, and pass to it another enum value or even an integer. It's definitely a quirk for people coming from modern languages.

Also, note that as it is today, each function receiving the enum as a parameter should call a validating function like `ensureValidEnum(theEnum)` to avoid repeating the check everywhere. So, it's boilerplate code which can be forgotten and certainly bite people thinking they are proper types.

Maybe the manual should make it more explicit and provide an example code:

```gdscript

func check_enum(value):
   match value:
       Enum.ONE,Enum.TWO,Enum.THREE:
            return value

   debug("Invalid enum value, return ONE as default")
   return Enum.ONE

func your_method(value1:Enum,value2:Enum):
    #Always sanitize your enums
    value1=check_enum(value1)
    value2=check_enum(value2)
    # ...
```

**Comment by vnen** [comment]
That is not true to C#.  You can assign any arbitrary number to an enum type, as long as cast it. If you have a `switch` statement with all valid enum values, it will still complain that not all code paths return a value.

Java is more strict on this and doesn't allow casting from int by default, but the `switch` statement still complains about not returning a value even if you cover all cases.

In summary, there's simply no way to guarantee the value is always valid at runtime (given GDScript is still dynamically typed by default) unless we check it everywhere, which will certainly affect performance. So in this situation, not unlike C# or Java, you need to handle the case in which the value falls outside the range.

> Also, note that as it is today, each function receiving the enum as a parameter should call a validating function like `ensureValidEnum(theEnum)` to avoid repeating the check everywhere.

The alternative being the engine calling something like this internally, like I mentioned above, which is a performance hit for everyone. You don't really **have** to call a function to validate, you can simply handle the cases where the enum may have a different value when needed.

```gdscript
func turn_left(dir)->Direction:
	match dir:
		Direction.LEFT:
			return Direction.DOWN
		Direction.RIGHT:
			return Direction.UP
		Direction.UP:
			return Direction.LEFT
		Direction.DOWN:
			return Direction.RIGHT
		_:
			assert(false, "Invalid Direction value")
			return Direction.UP # Just to return something.
```

Surely it is not the most convenient, but it is not the end of the world either.

**Comment by programaths** [comment]
I also mentionned Kotlin, which is high performance and does check all code path when using an enum  (or sealed types) (https://github.com/godotengine/godot/issues/71763#issuecomment-1399229592)

Here is a link to the description of "when" in Kotlin: https://kotlinlang.org/docs/control-flow.html#when-expression

The huge issue is that the programmer has to do some in place validation and when the enum is updated (new value added), the game would still run kinda fine and good luck to find where to add the new value in your code.

And in fact, even in Java, such error already popped in production, because programmers can't consider all the uses. So they search for references and sometimes miss one.

A la Kotlin, the app will just not complain and if a lazy dev wrote an "else", then it's his fault if he is in troubles later!

As I understand, enums should not be used too much outside the enclosing class, except as parameters. With the gotcha that the calee has to do some in place validation.

It could be a nice feature to warn the user that enums are improper types. ("Enum types are aliases of int")

I am probably too spoiled with Kotlin ^^

**Comment by vnen** [comment]
I understand all that, but given GDScript is still dynamically typed by default, it's very difficult to enforce enums to always be valid. It would require runtime checks that can become quite expensive, it's just not deemed worthy for a language like GDScript. This is a design decision backed by the technical limitations.

Note that some users actually use this feature of `enum` being equivalent to `int`. There would certainly be complaints if this restriction were added. 

We could consider a new type that works like `enum` but does guarantee the value to be in range by validating at runtime (or perhaps an annotation for `enum`). This way users would (or at least should) be aware of the performance implications by such type. This, however, should be posted as a [proposal](https://github.com/godotengine/godot-proposals) and certainly won't be done in time for 4.0.

**Comment by anvilfolk** [comment]
If what you are saying is an error, I wouldn't mind the error becoming a _warning_ when the enum has been fully checked within the `match` statement. The example would definitely need the `dir` parameter to be typed as the enum though :)

But I also think the GDScript team has been going in the direction of enums just being ints, and if that's the case then I do believe this is intended behavior.

**Comment by programaths** [comment]
That could be checked at editing time and be a simple warning, with an option to disable that check per project.

That way, no performance issue at runtime!

The whole point is to help tge developer in handling new cases everywhere without having to play seek and hunt through the codebass.

In IntelliJ, they have inspections that even make that work in Java. So, Java will not stop compiling because you didn't handle all cases, but the IDE will warn you about it.

I know, it sounds stupid, but it's nice when you add a document type in an enum and the IDE tells you all the places where you should handle that new case.

I found Kotlin better with refusing to compile, because that means that tge developer is required to actively think about, but that's a matter of "taste" here.

Having the check only at "editing" time is already very powerfull. 

Godot already does that for unused variables!

**Comment by anvilfolk** [comment]
So I'm starting to get the feeling that this is would be a useful _feature_ - instead of considering current behavior a bug.

Maybe adding an annotation to `match` for enums would be a good solution:

```GDScript
enum Named {V1, V2}

func enum_return(x : Named) -> int: 
  @warn_incomplete_enum
  match x:
    Named.V1: return 1
    Named.V2: return 2

  return 0
```

So this would not cause any issues. But if you changed it to `enum Named {V1, V2, V3}`, then it would. There's quite a few details to discuss here though. What happens if you have a `default` in your `match`, whether this ought to be `@dont_warn_incomplete_enum` and just have warnings by default, etc etc.

**Comment by programaths** [comment]
The annotation would indeed be more granular than a setting at project level!
But would not be less discoverable. Always pesky tradeoffs!

If one would use the annotation `@warn_incomplete_enum`, then using `default` would have to be a mistake, because the two are against each other.

As for adding warnings as default, then it would become too disruptive as it would popup in everyone codebase.

I think `@require_complete_enum`, which would prevent compilation, would be much better. So, big fat error with refusal to proceed. Really forcing the developer to account for new cases. 

As it would be an opt in which requires some effort (as the initial dev has to type the annotation), it would only be used when it's critical. So, the IDE and compiler can be loud and strong without issue!

**Comment by anvilfolk** [comment]
@akien-mga should we tag this as an enhancement for 4.x?

**Comment by YuriSizov** [comment]
@anvilfolk We don't have to assign a milestone to an issue, but if you think this is something we will be implementing, then sure, can be assigned 4.x.

**Comment by akien-mga** [comment]
I think this is more proposal material, no?
